cantocrabs/CantoCasino.sol

```diff
@@ -0,0 +1,263 @@
+        uint256 payout
+    );
+
+    uint256 _deployDate;
+
+    constructor(address _nftAddress, uint256 _maxBet)
+        Auth(msg.sender)
+        NFTRevenue(_nftAddress)
+    {
+        maxBet = _maxBet;
+        _deployDate = block.timestamp;
+        _houseAddress = msg.sender;
+    }
+
+    event Received(address, uint256);
+
+    receive() external payable {}
+
+    function play(
+        uint256 choice,
+        uint256 betType,
+        uint256 referrerId
+    )
+        public
+        payable
+        isValidRollNumber(choice)
+        isValidBetType(betType)
+        nonReentrant
+    {
+        require(msg.value <= maxBet, "Bet exceeds max betting limit");
+        require(
+            choice <= PAYOUT[betType],
+            "Choice is not compatible with the bet type selected"
+        );
+        require(
+            playerToBetId[msg.sender] == 0,
+            "User already has bet placed. Resolve said bet"
+        );
+
+        if (
+            playerToPlayerId[msg.sender] == 0 &&
+            !hasReferrer(msg.sender) &&
+            referrerId != 0 &&
+            referrerId <= playerCount
+        ) {
+            address _referrer = playerIdToPlayer[referrerId];
+            setupReferent(msg.sender, _referrer);
+        }
+
+        uint256 nftAmount = msg.value.mul(_nftFee).div(FEE_DENOM);
+
+        uint256 houseAmount = msg.value.mul(_houseEdge).div(FEE_DENOM);
+
+        uint256 referralAmount = msg.value.mul(_referralFee).div(FEE_DENOM);
+
+        uint256 finalBetAmount = msg.value.sub(nftAmount).sub(houseAmount).sub(
+            referralAmount
+        );
+
+        betsCounter.increment();
+
+        uint256 currentBetId = betsCounter.current();
+
+        Bet memory betData = Bet(
+            currentBetId,
+            finalBetAmount,
+            choice,
+            betType,
+            msg.sender,
+            GameStatus.PENDING
+        );
+
+        playerToBetId[msg.sender] = currentBetId;
+
+        bets[currentBetId] = betData;
+
+        if (playerToPlayerId[msg.sender] == 0) {
+            playerCount += 1;
+            playerToPlayerId[msg.sender] = playerCount;
+            playerIdToPlayer[playerCount] = msg.sender;
+        }
+
+        if (hasReferrer(msg.sender)) {
+            payRefferal(msg.sender, referralAmount);
+        } else {
+            payHouse(referralAmount);
+        }
+
+        payHouse(houseAmount);
+
+        updateNFTPool(nftAmount);
+
+        emit BetPlaced(currentBetId, msg.sender, msg.value, choice, betType);
+    }
+
+    function resolve() public nonReentrant {
+        require(
+            playerToBetId[msg.sender] != 0,
+            "No unresolved bets for this player"
+        );
+        require(
+            bets[playerToBetId[msg.sender]].status == GameStatus.PENDING,
+            "This bet is not pending"
+        );
+
+        uint256 previousBetId = playerToBetId[msg.sender];
+
+        //TODO: GET RANDOM NUMBER AND ANALYZE THE ANSWER

```

Comment:
Great work! [Nexus Oracles](https://github.com/NexusOracles) would love to provide randomness here!

Reply:
Sure, would love to work with you guys! Contact us in our Discord or send us an email: contact@cantocrabs.com.